src/panel/pages/timeline/components/TimelineTooltip.tsx

```diff
@@ -42,59 +55,103 @@ const TooltipElement = styled.p<{ position?: TooltipPosition }>`
     border-left: 6px solid transparent;
     border-right: 6px solid transparent;
     margin-top: -1px;
-    left: 50%;
+    left: calc(50% - ${(p) => p.offset}px);
     top: 100%;
     transform: translate(-50%, 0);
   }
 `;
 
 export const useTooltip = () => {
-  const ref: MutableRefObject<HTMLElement | null> = useRef<HTMLElement>(null);
+  const containerRef: MutableRefObject<HTMLElement | null> = useRef<
+    HTMLElement
+  >(null);
+  const tooltipRef: MutableRefObject<HTMLElement | null> = useRef<HTMLElement>(
+    null
+  );
+  const tooltipOffset: MutableRefObject<number> = useRef(0);
+  const mouseX = useRef<number | undefined>(undefined);
   const [hasRef, setHasRef] = useState(false);
   const [isVisible, setIsVisible] = useState(false);
   const [tooltipProps, setTooltipProps] = useState<
-    ComponentProps<typeof TimelineTooltip>
+    ComponentProps<typeof TimelineTooltip> & TimelineTooltipProps
   >({});
 
   const calculateTooltipPosition = useCallback(() => {
-    if (!ref.current) {
+    if (!containerRef.current) {
       return;
     }
 
-    const { x, y, width } = ref.current.getBoundingClientRect(); //eslint-disable-line @typescript-eslint/no-non-null-assertion
-    setTooltipProps({
+    const { x, y, width } = containerRef.current.getBoundingClientRect(); //eslint-disable-line @typescript-eslint/no-non-null-assertion
+
+    const newTooltipProps: ComponentProps<typeof TimelineTooltip> &
+      TimelineTooltipProps = {
       style: {
         position: "fixed",
         left: x + width / 2,
         top: y,
         transform: `translateX(-50%) translateY(-100%)`,
+        paddingBottom: 8,
+        zIndex: 9,
       },
+      handleTooltipRef,
+      tooltipOffset: tooltipOffset.current,
+    };
+
+    if (!tooltipRef.current || !!tooltipOffset.current) {
+      setTooltipProps(newTooltipProps);
+      return;
+    }
+
+    const offsetRight =
+      window.innerWidth - tooltipRef.current.getBoundingClientRect().right;
+    const offsetLeft = tooltipRef.current.getBoundingClientRect().left;
+
+    // * add an offset to the tooltip if it would be rendered out of bounds
+    if (offsetLeft < 0) {
+      tooltipOffset.current = Math.abs(offsetLeft) + 8;
```

Comment:
As tooltipOffset looks to be stateful, I think we can remove the ref?

```js
 const calculateTooltipPosition = useCallback(() => {
    if (!containerRef.current) {
      return;
    }

    const { x, y, width } = containerRef.current.getBoundingClientRect();

    const props = {
      style: {
        position: "fixed",
        left: x + width / 2,
        top: y,
        transform: `translateX(-50%) translateY(-100%)`,
        paddingBottom: 8,
      },
      handleTooltipRef,
    };

    if (!tooltipRef.current) {
      return setTooltipProps(props);
    }

    const offsetRight =
      window.innerWidth - tooltipRef.current.getBoundingClientRect().right;
    const offsetLeft = tooltipRef.current.getBoundingClientRect().left;

    // * add an offset to the tooltip if it would be rendered out of bounds
    if (offsetLeft < 0) {
      return setTooltipProps({
        ...props,
        tooltipOffset: Math.abs(offsetLeft) + 8,
      });
    }

    if (offsetRight < 0) {
      return setTooltipProps({
        ...props,
        tooltipOffset: offsetRight - 8,
      });
    }

    setTooltipProps({
      ...newTooltipProps,
      tooltipOffset: 0,
    });
  }, []);
```

Reply:
This doesn't seem to work sadly. The [condition on the ref](https://github.com/FormidableLabs/urql-devtools/blob/0e0d44a8cecf70d460dc976b3f510138082b7ba3/src/panel/pages/timeline/components/TimelineTooltip.tsx#L100) is needed to stop a weird double render from happening

Reply:
Have added a corresponding comment